### DISCUSSION TITLE
How to use WEBGPU with sPROXY_TO_PTHREAD and sOFFSCREENCANVAS_SUPPORT

### DISCUSSION THREAD

**Discussion by alienself** [emscripten-core/emscripten#19225] [General]
Hi everyone

I am wondering how to run webgpu off the main thread in a separate thread while rendering on an offscreen canvas passed from the main thread.
Do I need to create a worker or does emscripten take care of everything?

I am using the `sPROXY_TO_PTHREAD` and `sOFFSCREENCANVAS_SUPPORT` flags.

Right now I have the following code in my main thread, however I get an error when trying to request the canvas context

```js
import appModule from '/app.js';

const canvas = document.querySelector("#canvas");

const module = await appModule ({
    onRuntimeInitialized() {
        console.log("app initilized.");
    },
    canvas,
});
```

```sh
DOMException: Failed to execute 'getContext' on 'HTMLCanvasElement': Cannot get context from a canvas that has transferred its control to offscreen. at _wgpuInstanceCreateSurface

var descriptorFromCanvasHTMLSelector = nextInChainPtr;
assert(descriptorFromCanvasHTMLSelector);
assert(GROWABLE_HEAP_U32()[descriptorFromCanvasHTMLSelector >>> 2] === 0);
var selectorPtr = GROWABLE_HEAP_U32()[descriptorFromCanvasHTMLSelector + 8 >>> 2];
assert(selectorPtr);
var canvas = findCanvasEventTarget(selectorPtr);
var context = canvas.getContext("webgpu"); <<< ERROR
assert(context);
```

I saw this related issue: https://github.com/emscripten-core/emscripten/issues/17776
but it does not make mention of proxy to pthread

Am I missing something here?

**Comment by kainino0x** [comment]
We don't have proxying support for WebGPU yet.
We had it for a little bit, but it turned out not to be complete enough to use. It's been reverted now and WebGPU will be executed in the same thread where you call the webgpu.h functions, at least for now.

**Reply by kainino0x** [reply]
See #19645
